### ISSUE TITLE
Issue with VR frustum culling

### ISSUE THREAD

**Issue by CesiumBen** [CesiumGS/cesium-unreal#559]
While in VR tiles get culled out based on head motion, due to multiple frustums from the camera and the sizing potentially being miscalculated. This occurs when VR "world to meters" is 100 (regular scale) or when it's larger as well. Adjusted world to meters units wouldn't change much but technically the eyes would be further apart in world units based on this scale.

The gif below displays it as best as possible, unfortunately the output window does not render the same extents as wearing a VR headset. The 1st half of the gif I'm far from the starting position and disappearing tiles occurs more frequently. While I return to the origin point (When I'm grabbing with the hands) the culling is much better and the tiles remain as they should.

![clipping](https://user-images.githubusercontent.com/79323980/126542594-20b739ce-322c-48a9-9104-892005928270.gif)

This test was made with updates from this [PR](https://github.com/CesiumGS/cesium-unreal/pull/547). Please advise on ideal testing conditions.

**Comment by javagl** [comment]
<sup>(I won't be able to help substantially here, ... the culling traversal is complex and nearly impossible to test, but as an attempt to be clear about the actual problem, even at the risk that others may find the answers to these questions "obvious":</sup>

The problem is that there are missing tiles like the one in the lower right in this frame, right? :

![MissingTile](https://user-images.githubusercontent.com/5597569/127156480-ceed892b-bbcc-48ae-8eb4-f0932b422f1d.png)

And (that may be important: ) These tiles are **only** missing for a **single** frame, right?
The crucial point is: I assume that when you keep the camera oriented as in the screenshot above, it **does** show the tile in the **next** frame, right?
If this was **not** the case, then the underlying issue might be a bug in the "bounding-box-vs-frustum" intersection tests. 
But if it *does* show the tile eventually, then this may be similar to https://github.com/CesiumGS/cesium-native/issues/186 , and it would be wothwhile to try out whether it still appears when disabling occlusion culling (see the (Cesium-Internal) https://udn.unrealengine.com/s/question/0D54z0000708yGxCAI/one-frame-delay-after-setvisibilitytrue-true for details)

**Comment by CesiumBen** [comment]
@javagl Thanks for the lead! The tiles aren't missing on a frame by frame basis, its purely by moving the camera. So they are being culled based on "incorrect" calculations. Basically tiles on the edge of the screen are not loading or are "missing".

The fact that only the right side in the example above has this issue I think would just help point to a possible fix, I've seen tiles cull out on the bottom (and top if you're underground, just pointing it out not that it's a requirement) and that's happened in multiple locations. The example above is unique and I'm not sure why it's occurring only on certain tiles on the right side.

**Comment by kring** [comment]
VR devices usually use off-center frustums, right? I bet that's the problem. Frustums are currently communicated from Unreal to cesium-native as simply: position, direction, up, viewport dimensions, and horizontal and vertical field of view angles. Nithin added support for multiple frustums like this, but it still assume the frustum is centered on the view direction.

**Comment by CesiumBen** [comment]
Yes the frustums are offset from each other by a certain amount, basically just like your eyes. Unfortunately for a few years now Unreal and Unity doesn't represent that as 2 cameras anymore, rather 1 camera that is "VR" so its hard to see exactly what happening under the hood, I'll post more as I see it.

The eye offset would also be modified by "World To Meters" in the world settings.

**Comment by kring** [comment]
Nah I don't mean the eyes are at different _locations_. I think that will already be handled by Nithin's changes. I mean that the frustum itself is asymmetrical. I'm totally not at all an expert on this and mostly just hand waving, but I think this page explains it reasonably well:
https://developer.oculus.com/blog/tech-note-asymmetric-field-of-view-faq/

**Comment by javagl** [comment]
It may appear to be nitpicking, but again, I think it is important in order to find the root of this issue:

> The tiles aren't missing on a frame by frame basis, its purely by moving the camera.

Imagine you were looking at the given scene, and then moved the camera into the configuration that I showed in the last screenshot (with the missing tile in the lower right), and then _immediately and totally stopped any camera movement_. Would the tile then appear after another frame, or would it *only* appear when you moved the camera a bit further?

The point is: The respective tile might be considered to be "occluded" in frame `X`, based on the current camera configuration. And it may become visible in frame `X+1`, due to a camera movement. But for the reasons discussed in detail in the linked UDN issue, it may happen that the tile then is still not rendered, because the state information that it was "occluded" is still cached. But if this is the cause, it will be rendered in frame `X+2` (even if the camera is not moved between the frames `X+1` and `X+2`).

**Comment by CesiumBen** [comment]
> Imagine you were looking at the given scene, and then moved the camera into the configuration that I showed in the last screenshot (with the missing tile in the lower right), and then immediately and totally stopped any camera movement. Would the tile then appear after another frame, or would it only appear when you moved the camera a bit further?

@javagl Thanks for diagnosing more!

I can stare an a space where a tile should be loaded but isn't for several seconds so I would say that it's a matter of the culling calculations. Sorry if the gifs above look rather quick, below is an example where I try to keep the head as steady as possible as I look upwards, my hand is waving to give you a sense of the framerate. You can see the tiles missing for more than a few frames. You'll notice some flickering, I believe that's because my head is on the edge of where the calculations consider it to be loaded or unloaded and my neck isn't quite as steady as a Komatsu excavator.

![VRCulling](https://user-images.githubusercontent.com/79323980/127394811-1aca2619-1a09-4668-8ee5-49c64ae69eef.gif)

_____
> Nah I don't mean the eyes are at different locations. I think that will already be handled by Nithin's changes. I mean that the frustum itself is asymmetrical. I think this page explains it reasonably well:
https://developer.oculus.com/blog/tech-note-asymmetric-field-of-view-faq/

@kring Great find! Not sure if this kinda thing is being implemented in Unreal at this current time but I'll follow up if I see anything more about it. From what I read it sounds like we might see more culling on one side than another if it is asymmetrical 

I went to Bolivia to the Salar de Uyuni (flattest place on earth) to do a different test, I'm assuming that occlusion wouldn't occur in this case but I noticed some interesting patterns on how it's culling only in certain directions. In the beginning of the gif you'll see that a lot of the tiles in the distance do not disapear at all, as I rotate to the right the tiles become prone to disappearing. Then as I go around some tiles remain persistent close to the user, but the ones further away do disappear. 

I couldn't tell if there was any asymmetry due to the inconsistent culling I saw in the test.

![FlatTest](https://user-images.githubusercontent.com/79323980/127398390-48862fe4-779c-4a7a-a85d-91f3462fa881.gif)


> I'm totally not at all an expert on this and mostly just hand waving, but 

Me too, I don't usually travel down the VR render pipeline to where the sausage is being made. 😛

**Comment by javagl** [comment]
The new GIFs stongly suggest that this is *not* an issue of occlusion culling (i.e. not something that is directly related to the Unreal Engine black box), but that there is *really* something wrong with our culling implementation.

This could be a bug in the "bounding-box-vs-frustum" intersection test (but this should be relatively simple, and thus unlikely). It's more likely that the frustum(s) for VR mode cannot be obtained in the same way as for the "normal" mode. All this is hard to test and guess. Wild googling brings things like https://answers.unrealengine.com/questions/243325/how-to-get-real-camera-position-in-vr-mode.html , but it's hard to know what to do with that...

<sup>(I still think that there should be "unit-test-tilesets" like a [subdivided plane](https://github.com/CesiumGS/cesium-native/issues/96#issuecomment-750345668) for things like this. But *if* somebody had set up a test for this a few months ago, he'd been told that he's setting the right priorities....)</sup>

**Comment by CesiumBen** [comment]
@nithinp7 Thank you for the updates! Good news and bad news

Good news first. The flat terrain test went perfectly. Every tile looked like it culled appropriately, or at least I could not see any tiles disappearing. 

![FlatTest2](https://user-images.githubusercontent.com/79323980/127888533-4cc6d950-bd0c-4dc4-bf0d-ddd836f8f54d.gif)

Bad news, a test on the mountains still had them disappearing. Unfortunately you can't see that in the gif below since the output to screen is a smaller FOV than the headset. I would guess this points us to the occlusion calculations. I would add that it feels like there is a little bit of an improvement.

![HillTest1](https://user-images.githubusercontent.com/79323980/127888830-c8082a15-536b-436b-97ab-03416c65dde2.gif)

**Comment by nithinp7** [comment]
@CesiumBen I have not yet been able to reproduce the latest version of the issue so far. I tried earlier in the week with HTC Vive and tried again just now with the Oculus Quest 2 + the same asset you're using. I'll continue looking for the issue in the code, but is there anything else particularly I can try to reproduce the problem on the headset?
